core/src/main/scala/com/phaller/rasync/Cell.scala

```diff
@@ -148,78 +216,120 @@ object Cell {
+
+private object IntermediateState {
+  def empty[K <: Key[V], V](updater: Updater[V]): IntermediateState[K, V] =
+    new IntermediateState[K, V](updater.initial, false, List(), Map(), Map(), Map(), Map())
+}
 
-  implicit val ctx = pool
+private class FinalState[K <: Key[V], V](
+  /** The final result of this cell.*/
+  val res: Try[V],
+  /** A list of cells that `this` cell depends on mapped to the callbacks to call, if those cells change. */
+  /*
+  * When a cell is completed, all `completeDependentCells` are copied from the IntermediateState
+  * to the FinalState. That way, we know that the dependentCell may poll a staged value (at most) one
+  * more time. After that, the dependentCell doese not depend on `this` cell any more and is removed
+  * from `completeDependentCells`. Repeated polls will return in NoOutcome.
+  */
+  val completeDependentCells: TrieMap[Cell[K, V], Unit],
+  /** A list of cells that `this` cell depends on mapped to the callbacks to call, if those cells change. */
+  val nextDependentCells: TrieMap[Cell[K, V], Unit])
+  extends State[V]
+
+/**
+ * Implementation of traits Cell and CellCompleter as the same run-time object.
+ *
+ * Instances of the class use a `State` to store the current value and dependency information.
+ */
+private class CellImpl[K <: Key[V], V](pool: HandlerPool, val key: K, updater: Updater[V], override val init: (Cell[K, V]) => Outcome[V]) extends Cell[K, V] with CellCompleter[K, V] {
+  implicit val ctx: HandlerPool = pool
 
-  private val nodepslatch = new CountDownLatch(1)
+  // Used for tests only.
+  private val nocompletedepslatch = new CountDownLatch(1)
+  // Used for tests only.
   private val nonextdepslatch = new CountDownLatch(1)
 
   /* Contains a value either of type
-   * (a) `Try[V]`      for the final result, or
-   * (b) `State[K,V]`  for an incomplete state.
-   *
-   * Assumes that dependencies need to be kept until a final result is known.
+   * (a) `FinalState[K, V]`         for the final result, or
+   * (b) `IntermediateState[K,V]`   for an incomplete state.
    */
-  private val state = new AtomicReference[AnyRef](State.empty[K, V](updater))
+  private val state = new AtomicReference[State[V]](IntermediateState.empty[K, V](updater))
+
+  // A list of callbacks to call, when `this` cell is completed/updated.
+  // Note that this is not sync'ed with the state in any way, so calling `onComplete` or `onNext`
+  // while tasks are running leads to non-deterministic calls.
+  private var onCompleteHandler: List[Try[V] => Any] = List()
+  private var onNextHandler: List[Try[V] => Any] = List()
 
   // `CellCompleter` and corresponding `Cell` are the same run-time object.
-  override def cell: Cell[K, V] = this
+  override val cell: Cell[K, V] = this
+  override val completer: CellCompleter[K, V] = this.asInstanceOf[CellCompleter[K, V]]
+
+  override def sequential[T](f: => T): T = this.synchronized(f)
```

Comment:
I'm not sure this is a good idea. The only thing that's achieved here is to hide the use of a monitor. I would argue this is a net loss, since the use of monitors has such dramatic consequences, both in terms of correctness (deadlocks) and in terms of performance.